pkg/cluster/fixetcd.go

```diff
@@ -0,0 +1,499 @@
+	jobName                             = "etcd-recovery-"
+	jobNameDataBackup                   = jobName + "data-backup"
+	jobNameFixPeers                     = jobName + "fix-peers"
+	patchOverides                       = "unsupportedConfigOverrides:"
+	patchDisableOverrides               = `{"useUnsupportedUnsafeNonHANonProductionUnstableEtcd": true}`
+	ctxKey                timeoutAdjust = "TESTING"
+)
+
+type timeoutAdjust string
+
+type degradedEtcd struct {
+	Node  string
+	Pod   string
+	NewIP string
+	OldIP string
+}
+
+func (m *manager) fixEtcd(ctx context.Context) error {
+	pods, err := m.kubernetescli.CoreV1().Pods(nameSpaceEtcds).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	// Exit early if degradedEtcd is empty, an error is returned
+	// no further actions should be taken
+	de, err := comparePodEnvToIP(m.log, pods)
+	if err != nil {
+		return err
+	}
+	m.log.Infof("Found degraded endpoint: %v", de)
+
+	err = backupEtcdData(ctx, m.log, de.Node, m.kubernetescli.BatchV1())
+	if err != nil {
+		return err
+	}
+
+	err = fixPeers(ctx, m.log, de, pods, m.kubernetescli, m.securitycli.SecurityV1().SecurityContextConstraints(), m.doc.OpenShiftCluster.Name)
+	if err != nil {
+		return err
+	}
+
+	etcd, err := m.operatorcli.OperatorV1().Etcds().Get(ctx, arov1alpha1.SingletonClusterName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	etcd.Spec.UnsupportedConfigOverrides = kruntime.RawExtension{
+		Raw: []byte(patchDisableOverrides),
+	}
+	err = patchEtcd(ctx, m.log, m.operatorcli.OperatorV1().Etcds(), etcd, patchDisableOverrides)
+	if err != nil {
+		return err
+	}
+
+	err = deleteSecrets(ctx, m.log, m.kubernetescli.CoreV1().Secrets(nameSpaceEtcds), de)
+	if err != nil {
+		return err
+	}
+
+	etcd.Spec.ForceRedeploymentReason = fmt.Sprintf("single-master-recovery-%s", time.Now())
+	err = patchEtcd(ctx, m.log, m.operatorcli.OperatorV1().Etcds(), etcd, etcd.Spec.ForceRedeploymentReason)
+	if err != nil {
+		return err
+	}
+
+	etcd.Spec.OperatorSpec.UnsupportedConfigOverrides.Reset()
+	return patchEtcd(ctx, m.log, m.operatorcli.OperatorV1().Etcds(), etcd, patchOverides+string(etcd.Spec.OperatorSpec.UnsupportedConfigOverrides.Raw))
+}
+
+func patchEtcd(ctx context.Context, log *logrus.Entry, operatercli operatorv1client.EtcdInterface, e *operatorv1.Etcd, patch string) error {
+	e.CreationTimestamp = metav1.Time{
+		Time: time.Now(),
+	}
+	e.ResourceVersion = ""
+	e.SelfLink = ""
+	e.UID = ""
+
+	buf := &bytes.Buffer{}
+	err := codec.NewEncoder(buf, &codec.JsonHandle{}).Encode(e)
+	if err != nil {
+		return err
+	}
+
+	_, err = operatercli.Patch(ctx, e.Name, types.MergePatchType, buf.Bytes(), metav1.PatchOptions{})
+	if err != nil {
+		return err
+	}
+	log.Infof("Patched etcd %s with %s", e.Name, patch)
+
+	return nil
+}
+
+func deleteSecrets(ctx context.Context, log *logrus.Entry, secretcli corev1client.SecretInterface, de *degradedEtcd) error {
+	secrets, err := secretcli.List(ctx, metav1.ListOptions{})
```

Comment:
It looks like we are only listing secrets in openshift-etcd namespace. It's passed in on line 94 where `deleteSecrets` is called. Maybe we should change that for readability, because here it looks like we're listing all secrets.

Regarding the regex matching happening here, If we can actually construct the exact names of the secrets we want to delete, without needing to list, we should do that. I'm seeing that they follow a format like "etcd-peer-"+clustername+infraID+nodeRole. The regex matching in the range might be error prone as Brendan said.

Reply:
@SudoBrendan These are secrets listed from namespace `openshift-etcd`. The names are generated, which is why I'm using regex to match the name. This should probably be improved, if the desired secrets aren't matched first I believe there's a chance others could be matched and deleted.

Reply:
I've tightened up the regex pattern to only match our desired three prefixes. I'm not sure how I can get the secret names without this, the secrets themselves don't have any labels or owners.